### PR TITLE
tests: Add note about test suite naming convention in developer-notes.md

### PR DESCRIPTION
**Pull request by practicalswift**
Changes:
* Add note about test suite naming convention
* Fix exceptions
* Add regression test

Rationale:
* Consistent naming of test suites makes programmatic test running of specific tests/subsets of tests easier
* Explicit is better than implicit

Before this commit:

```
$ contrib/devtools/lint-tests.sh
The test suite in file src/test/foo_tests.cpp should be named
"foo_tests". Please make sure the following test suites follow
that convention:

src/test/blockchain_tests.cpp:BOOST_FIXTURE_TEST_SUITE(blockchain_difficulty_tests, BasicTestingSetup)
src/test/prevector_tests.cpp:BOOST_FIXTURE_TEST_SUITE(PrevectorTests, TestingSetup)
src/wallet/test/coinselector_tests.cpp:BOOST_FIXTURE_TEST_SUITE(coin_selection_tests, WalletTestingSetup)
src/wallet/test/crypto_tests.cpp:BOOST_FIXTURE_TEST_SUITE(wallet_crypto, BasicTestingSetup)
$
```

After this commit:

```
$ contrib/devtools/lint-tests.sh
$
```

### CODE REVIEW

**Review by randolf — APPROVED** [review]

File: contrib/devtools/lint-tests.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+#
+# Copyright (c) 2018 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#
+# Check the test suite naming convention
+
+NAMING_INCONSISTENCIES=$(git grep -E '^BOOST_FIXTURE_TEST_SUITE\(' -- \
+    "src/test/**.cpp" "src/wallet/test/**.cpp" | \
```
Comment (conscott) [inline contrib/devtools/lint-tests.sh]: There is also a violation in 
```
src/univalue/test/object.cpp:BOOST_FIXTURE_TEST_SUITE(univalue_tests, BasicTestingSetup)
```
but I realize that is outside the scope of `test_bitcoin` tests. 

Coule make changes upstream to `univalue` as well, but that's minor.

Comment (practicalswift) [reply contrib/devtools/lint-tests.sh]: Yes, upstreams' tests are intentionally not checked by `lint-tests.sh` since we don't want Travis to block on things where we need help from upstreams. All violations should be possible to fix in-tree :-)

**Comment by conscott** [comment]
Tested ACK db983beba6fec910722bacc0d67c7ae98ba566fa

**Comment by eklitzke** [comment]
Since this fixes all violations of the rule we should also add this a a required lint check in Travis.

File: contrib/devtools/lint-tests.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    grep -vE '/(.*?)\.cpp:BOOST_FIXTURE_TEST_SUITE\(\1, .*\)$')
+if [[ ${NAMING_INCONSISTENCIES} != "" ]]; then
```
Comment (eklitzke) [inline contrib/devtools/lint-tests.sh]: There is a serious shell quoting problem here. This isn't just a nit, it changes the semantics of the test. Use test -n and quote the variable, e.g.

```bash
if [ -n "${NAMING_INCONSISTENCIES}" ]; then
    ...
```

**Review by eklitzke — CHANGES_REQUESTED** [review]

File: contrib/devtools/lint-tests.sh (hunk 2 above)
Comment (eklitzke) [reply contrib/devtools/lint-tests.sh]: Actually I have to take this back, apparently [[ doesn't perform word-splitting. Still scares me though.

**Review by eklitzke — APPROVED** [review]
tested ACK db983beba, but would like to see this added to the travis tests

File: contrib/devtools/lint-tests.sh (hunk 2 above)
Comment (practicalswift) [reply contrib/devtools/lint-tests.sh]: @eklitzke `shellcheck` seems happy :-)

**Comment by practicalswift** [comment]
@eklitzke Scripts matching `contrib/devtools/lint-*.sh` are automatically run by Travis, so no need to add it explicitly :-)

File: contrib/devtools/lint-tests.sh (hunk 2 above)
Comment (practicalswift) [reply contrib/devtools/lint-tests.sh]: No need to be scared :-)  `[[ ${FOO} != "" ]]` is perfectly safe AFAIK. `shellcheck` agrees :-)

**Comment by MarcoFalke** [comment]
utACK db983beba6fec910722bacc0d67c7ae98ba566fa

File: contrib/devtools/lint-tests.sh (hunk 2 above)
Comment (practicalswift) [reply contrib/devtools/lint-tests.sh]: This is a `bash` script and is explicit about it (`#!/bin/bash`), so no need for follow-up PR:s for other shells :-)
